File: external_apps/mobile_ios/mobile_ios/Features/Payment/PaymentViewModel.swift
//
//  PaymentViewModel.swift
//  mobile_ios
//
//  Created for Payment & Billing Feature
//

import Foundation

@MainActor
class PaymentViewModel: ObservableObject {
    
    @Published var historyState: ViewState<[Transaction]> = .idle
    @Published var stripeConfigState: ViewState<StripePaymentConfig> = .idle
    @Published var paymentUrl: URL?
    @Published var invoiceState: ViewState<Invoice> = .idle
    
    // In a real app, inject this via DI
    private let repository: PaymentRepositoryProtocol
    let storeKitManager = StoreKitManager()
    
    init(repository: PaymentRepositoryProtocol = MockPaymentRepository()) {
        self.repository = repository
        fetchHistory()
    }

    func purchaseSubscription() {
        Task {
            if let product = await storeKitManager.products.first {
                try? await storeKitManager.purchase(product)
            } else {
                print("No products available")
            }
        }
    }
    
    func fetchHistory() {
        historyState = .loading
        Task {
            do {
                let transactions = try await repository.getTransactionHistory()
                DispatchQueue.main.async {
                    self.historyState = .success(transactions)
                }
            } catch {
                DispatchQueue.main.async {
                    self.historyState = .error(error.localizedDescription)
                }
            }
        }
    }
    
    func prepareStripePayment() {
        stripeConfigState = .loading
        Task {
            do {
                let config = try await repository.getStripePaymentConfig()
                DispatchQueue.main.async {
                    self.stripeConfigState = .success(config)
                }
            } catch {
                DispatchQueue.main.async {
                    self.stripeConfigState = .error(error.localizedDescription)
                }
            }
        }
    }
    
    func initiatePayment(method: String, amount: Double) {
        let urlString = "https://example.com/pay?method=\(method)&amount=\(amount)"
        if let url = URL(string: urlString) {
            DispatchQueue.main.async {
                self.paymentUrl = url
            }
        }
    }
    
    func fetchInvoice(transactionId: String) {
        invoiceState = .loading
        Task {
            do {
                let invoice = try await repository.getInvoice(transactionId: transactionId)
                DispatchQueue.main.async {
                    self.invoiceState = .success(invoice)
                }
            } catch {
                DispatchQueue.main.async {
                    self.invoiceState = .error(error.localizedDescription)
                }
            }
        }
    }
}

File: external_apps/mobile_android/app/src/main/java/com/example/mobile_android/ui/payment/PaymentViewModel.kt
package com.example.mobile_android.ui.payment

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.mobile_android.data.model.Transaction
import com.example.mobile_android.data.repository.PaymentRepository
import com.example.mobile_android.data.model.StripePaymentConfig
import com.example.mobile_android.ui.core.UiState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject
import kotlinx.coroutines.delay

import com.example.mobile_android.data.billing.BillingManager
import android.app.Activity

@HiltViewModel
class PaymentViewModel @Inject constructor(
    private val repository: PaymentRepository,
    private val billingManager: BillingManager
) : ViewModel() {

    private val _historyState = MutableStateFlow<UiState<List<Transaction>>>(UiState.Loading)
    val historyState: StateFlow<UiState<List<Transaction>>> = _historyState.asStateFlow()

    init {
        billingManager.startConnection()
        fetchHistory()
    }



    private val _stripeConfigState = MutableStateFlow<UiState<StripePaymentConfig>?>(null)
    val stripeConfigState: StateFlow<UiState<StripePaymentConfig>?> = _stripeConfigState.asStateFlow()

    fun fetchHistory() {
        viewModelScope.launch {
            _historyState.value = UiState.Loading
            val result = repository.getTransactionHistory()
            result.onSuccess {
                _historyState.value = UiState.Success(it)
            }.onFailure {
                _historyState.value = UiState.Error(it.message ?: "Unknown error")
            }
        }
    }

    fun prepareStripePayment() {
        viewModelScope.launch {
            _stripeConfigState.value = UiState.Loading
            val result = repository.getStripePaymentConfig()
            result.onSuccess {
                _stripeConfigState.value = UiState.Success(it)
            }.onFailure {
                _stripeConfigState.value = UiState.Error(it.message ?: "Failed to fetch config")
            }
        }
    }

    private val _paymentUrlState = MutableStateFlow<String?>(null)
    val paymentUrlState: StateFlow<String?> = _paymentUrlState.asStateFlow()

    fun initiatePayment(method: String, amount: Double) {
        viewModelScope.launch {
            // Mock: Get URL from repository based on method
            // In real app: response = repository.createPaymentIntent(method, amount, "VND")
            // if success -> _paymentUrlState.value = response.url
            delay(500)
            _paymentUrlState.value = "https://example.com/pay?method=$method&amount=$amount"
        }
    }
    
    fun onPaymentUrlOpened() {
        _paymentUrlState.value = null
    }

    fun onPaymentResult(success: Boolean) {
        if (success) {
            fetchHistory() // Refresh history
        }
    }
    
    fun subscribe(activity: Activity) {
        billingManager.launchPurchaseFlow(activity)
    }
    
    private val _invoiceState = MutableStateFlow<UiState<com.example.mobile_android.data.model.Invoice>?>(null)
    val invoiceState: StateFlow<UiState<com.example.mobile_android.data.model.Invoice>?> = _invoiceState.asStateFlow()
    
    fun fetchInvoice(transactionId: String) {
        viewModelScope.launch {
            _invoiceState.value = UiState.Loading
            val result = repository.getInvoice(transactionId)
            result.onSuccess {
                _invoiceState.value = UiState.Success(it)
            }.onFailure {
                _invoiceState.value = UiState.Error(it.message ?: "Failed to fetch invoice")
            }
        }
    }
}
